### ISSUE TITLE
group.election().term().leader() returns null for non leading group members

### ISSUE THREAD

**Issue by cachenbach** [atomix/atomix#186]
Hello and thank you very much for this great framework!

I have a problem when a _non_ leading member of a group wants to know who the leader is: `group.election().term().leader();`
The method is always returning `null` even if a leader is present. Is this the correct behavior?
Only the leader himself is getting the right answer: LocalGroupMember[id=...]

I am using: 1.0.0-rc4

Thanks for any help!
 Christian

**Comment by kuujo** [comment]
This is *not* expected behavior. I will try to reproduce this.

**Comment by cachenbach** [comment]
Perhaps I can help with some logs:

First client:

`11:29:48 396  DEBUG io.atomix.copycat.client.util.ClientConnection - Setting up connection to localhost/127.0.0.1:8700
11:29:48 396  DEBUG io.atomix.copycat.client.util.ClientConnection - Sending ConnectRequest[client=9ec58c33-042d-4db6-a653-59a561f92f90]
11:29:48 410  DEBUG io.atomix.copycat.client.util.ClientConnection - Received ConnectResponse[status=OK, leader=localhost/127.0.0.1:8700, members=[localhost/127.0.0.1:8700]]
11:29:48 418  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received CommandResponse[status=OK, index=140116, eventIndex=140114, result=11595]
11:29:48 420  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received KeepAliveResponse[status=OK, error=null, leader=localhost/127.0.0.1:8700, members=[localhost/127.0.0.1:8700]]
11:29:48 422  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Sending QueryRequest[session=140114, sequence=1, index=140116, query=InstanceQuery[resource=11595, query=io.atomix.resource.internal.ResourceCommand$Config@24853a22, consistency=null]]
11:29:48 433  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received QueryResponse[status=OK, index=140117, eventIndex=140114, result={}]
11:29:48 440  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Sending CommandRequest[session=140114, sequence=2, command=io.atomix.manager.internal.GetResource@3a36046a]
11:29:48 443  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received CommandResponse[status=OK, index=140118, eventIndex=140114, result=111618]
11:29:48 448  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Sending CommandRequest[session=140114, sequence=3, command=InstanceCommand[resource=111618, command=ResourceCommand[command=io.atomix.group.internal.GroupCommands$Listen@7ce7c981]]]
11:29:48 451  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received CommandResponse[status=OK, index=140119, eventIndex=140114, result=[]]
11:29:48 452  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Sending CommandRequest[session=140114, sequence=4, command=InstanceCommand[resource=111618, command=ResourceCommand[command=io.atomix.group.internal.GroupCommands$Join@151203eb]]]
11:29:48 462  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received PublishRequest[session=140114, eventIndex=140120, previousIndex=140114, events=[Event[event=join, message=InstanceEvent[resource=111618, message=GroupMemberInfo[member=ae1ebebe-d4c7-4035-84f4-b484455785bd]]], Event[event=elect, message=InstanceEvent[resource=111618, message=ae1ebebe-d4c7-4035-84f4-b484455785bd]]]]
11:29:48 465  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Received CommandResponse[status=OK, index=140120, eventIndex=140114, result=GroupMemberInfo[member=ae1ebebe-d4c7-4035-84f4-b484455785bd]]
11:29:48 467  DEBUG io.atomix.copycat.client.session.ClientSession - 140114 - Sending CommandRequest[session=140114, sequence=5, command=InstanceCommand[resource=111618, command=ResourceCommand[command=io.atomix.group.internal.GroupCommands$Listen@774d7723]]]

Joined as: ae1ebebe-d4c7-4035-84f4-b484455785bd

Leader: LocalGroupMember[id=ae1ebebe-d4c7-4035-84f4-b484455785bd]
`

Second client:

``11:29:52 375  DEBUG io.atomix.copycat.client.util.ClientConnection - Setting up connection to localhost/127.0.0.1:8700
11:29:52 375  DEBUG io.atomix.copycat.client.util.ClientConnection - Sending ConnectRequest[client=436b8c82-be4b-413c-9df3-1154887ae8dd]
11:29:52 391  DEBUG io.atomix.copycat.client.util.ClientConnection - Received ConnectResponse[status=OK, leader=localhost/127.0.0.1:8700, members=[localhost/127.0.0.1:8700]]
11:29:52 397  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received CommandResponse[status=OK, index=140128, eventIndex=140126, result=11595]
11:29:52 398  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received KeepAliveResponse[status=OK, error=null, leader=localhost/127.0.0.1:8700, members=[localhost/127.0.0.1:8700]]
11:29:52 401  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Sending QueryRequest[session=140126, sequence=1, index=140128, query=InstanceQuery[resource=11595, query=io.atomix.resource.internal.ResourceCommand$Config@5eeabbd6, consistency=null]]
11:29:52 405  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received QueryResponse[status=OK, index=140129, eventIndex=140126, result={}]
11:29:52 412  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Sending CommandRequest[session=140126, sequence=2, command=io.atomix.manager.internal.GetResource@78c90822]
11:29:52 415  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received CommandResponse[status=OK, index=140130, eventIndex=140126, result=111618]
11:29:52 419  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Sending CommandRequest[session=140126, sequence=3, command=InstanceCommand[resource=111618, command=ResourceCommand[command=io.atomix.group.internal.GroupCommands$Listen@ac5e11a]]]
11:29:52 423  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received CommandResponse[status=OK, index=140131, eventIndex=140126, result=[GroupMemberInfo[member=ae1ebebe-d4c7-4035-84f4-b484455785bd]]]
11:29:52 424  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Sending CommandRequest[session=140126, sequence=4, command=InstanceCommand[resource=111618, command=ResourceCommand[command=io.atomix.group.internal.GroupCommands$Join@55dffedb]]]
11:29:52 435  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received PublishRequest[session=140126, eventIndex=140132, previousIndex=140126, events=[Event[event=join, message=InstanceEvent[resource=111618, message=GroupMemberInfo[member=9717001f-a1ed-4390-812b-30a8df51beb0]]]]]
11:29:52 439  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Received CommandResponse[status=OK, index=140132, eventIndex=140126, result=GroupMemberInfo[member=9717001f-a1ed-4390-812b-30a8df51beb0]]
11:29:52 442  DEBUG io.atomix.copycat.client.session.ClientSession - 140126 - Sending CommandRequest[session=140126, sequence=5, command=InstanceCommand[resource=111618, command=ResourceCommand[command=io.atomix.group.internal.GroupCommands$Listen@438ce31f]]]

Joined as: 9717001f-a1ed-4390-812b-30a8df51beb0

Leader: null
``

**Comment by kuujo** [comment]
@cachenbach I haven't been able to reproduce this, and I should probably restate my answer as to whether this is expected behavior. Whether this behavior is expected depends on the specific code. There's no guarantee that once node A becomes a leader, node B will see that leader. There's only a guarantee that both nodes A and B will see term and leader changes occur in the same order (sequential consistency). To ensure node B will see node A, node B must wait for an election event:

```java
group.election().onElection(term -> {
  // term.leader() should be non-null
});
```

When a node starts, there's no guarantee `group.election().term().leader()` will be non-null until that election event is received.

**Comment by cachenbach** [comment]
Thank you for the clarification!

My initial intention was to send a message directly to the leader. To wait for an election is perhaps not a good idea. Is it possible to trigger an election?  

Thanks!

**Comment by kuujo** [comment]
So, I did discover this bug, and it's fixed in #195
